flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/source/reader/ArrayBatchRecords.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * This class uses array to store a batch of records from the same file (with the same fileOffset).
+ */
+class ArrayBatchRecords<T> implements RecordsWithSplitIds<RecordAndPosition<T>> {
+  @Nullable
+  private String splitId;
+  @Nullable
+  private final Pool.Recycler<T[]> recycler;
+  @Nullable
+  private final T[] records;
+  private final int numberOfRecords;
+  private final Set<String> finishedSplits;
+  private final RecordAndPosition<T> recordAndPosition;
+
+  // point to current read position within the records array
+  private int position;
+
+  private ArrayBatchRecords(
+      @Nullable String splitId, @Nullable Pool.Recycler<T[]> recycler, @Nullable T[] records,
+      int numberOfRecords, int fileOffset, long startingRecordOffset, Set<String> finishedSplits) {
+    Preconditions.checkArgument(numberOfRecords >= 0, "numberOfRecords can't be negative");
+    Preconditions.checkArgument(fileOffset >= 0, "fileOffset can't be negative");
+    Preconditions.checkArgument(startingRecordOffset >= 0, "numberOfRecords can't be negative");
+
+    this.splitId = splitId;
+    this.recycler = recycler;
+    this.records = records;
+    this.numberOfRecords = numberOfRecords;
+    this.finishedSplits = Preconditions.checkNotNull(finishedSplits, "finishedSplits can be empty but not null");
+    this.recordAndPosition = new RecordAndPosition<>();
+
+    recordAndPosition.set(null, fileOffset, startingRecordOffset);
+    this.position = 0;
+  }
+
+  @Nullable
+  @Override
+  public String nextSplit() {
+    String nextSplit = this.splitId;
+    // set the splitId to null to indicate no more splits
+    // this class only contains record for one split
+    this.splitId = null;
+    return nextSplit;
+  }
+
+  @Nullable
+  @Override
+  public RecordAndPosition<T> nextRecordFromSplit() {
+    if (position < numberOfRecords) {
+      recordAndPosition.record(records[position++]);
+      return recordAndPosition;
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * This method is called when all records from this batch has been emitted.
+   * If recycler is set, it should be called to return the records array back to pool.
+   */
+  @Override
+  public void recycle() {
+    if (recycler != null) {
+      recycler.recycle(records);
+    }
+  }
+
+  @Override
+  public Set<String> finishedSplits() {
+    return finishedSplits;
+  }
+
+  @VisibleForTesting
+  T[] records() {
+    return records;
+  }
+
+  @VisibleForTesting
+  int numberOfRecords() {
+    return numberOfRecords;
+  }
+
+  /**
+   * Create a ArrayBatchRecords backed up an array with records from the same file
+   *
+   * @param splitId Iceberg source only read from one split a time.
+   *                We never have multiple records from multiple splits.
+   * @param recycler Because {@link DataIterator} with {@link RowData} returns an iterator of reused RowData object,
+   *                 we need to clone RowData eagerly when constructing a batch fo records.
```

Comment:
Nit: Typo `fo` instead of `of` in `batch of records`.

Reply:
good batch. fixed